Allow vaulted string to be used as actual strings (#73054)

eqrx opened the pull request:
##### SUMMARY
Multiple places in the code base expect actual `str` types and
refuse to work with objects that can be treated as strings.
This commit lets `AnsibleVaultEncryptedUnicode` deceive those
places by faking to be a `str`.

Fixes #73054

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
`AnsibleVaultEncryptedUnicode`

bcoca commented on lib/ansible/parsing/yaml/objects.py:
```diff
@@ -113,6 +113,9 @@ def __init__(self, ciphertext):
         self._ciphertext = to_bytes(ciphertext)
 
+    __class__ = text_type
+
```
shouldn't we just inherit from AnsibleUnicode instead of AnsibleBaseYAMLObject

webknjaz replied:
Yeah, this looks dangerous. But also inheriting from string is rather problematic. I recall attempting this a year ago in our migration script and it was rather painful.

eqrx replied:
that is smarter :) agreed

eqrx commented:
@bcoca looking at the failed test it seems that the change indeed fixes the direct bug, but lets other tests fail because the value gets serialized to another representation. What do you think would be the best approach? Would multi inheritance solve this problem without side effects, should I adapt the tests (sounds like a bad idea) or should I dig down more into the problem and try to find another solution (will do that until you answer :) )? Thank

bcoca commented:
` serialized to another representation` this should always be the case with vaults .. its basically how it 'unvaults'

eqrx commented:
(fyi, I working on this. Just taking longer since I need to take a closer look at the testing system)

bcoca commented on lib/ansible/module_utils/common/json.py:
```diff
@@ -49,6 +49,15 @@ def __init__(self, preprocess_unsafe=False, vault_to_text=False, **kwargs):
+
     # NOTE: ALWAYS inform AWS/Tower when new items get added as they consume them downstream via a callback
     def default(self, o):
         if getattr(o, '__ENCRYPTED__', False):
```
do we still need this entry?

eqrx replied:
well, if I leave this entry out the json encoder just encodes it as a string without consulting the `default` method, which breaks everything

bcoca replied:
i mean the entry in the default that check encrypted, now that it is in encode, it should never get there.

eqrx replied:
oh sorry, jup

Commit:
AnsibleVaultEncryptedUnicode is now a string (#73054)

Multiple places in the code base expect actual `str` types and
refuse to work with objects that can be treated as strings.
This commit lets `AnsibleVaultEncryptedUnicode` fake this by
having a `str` value as `__class__`.

mkrizek commented:
!waffling core_review

eqrx commented:
My solution does not work, closing
